## imort library


In [3]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
import os

In [7]:
PROJ_DIR=r"/home/sunbeam/Documents/flight_price"
DATA_DIR="data"


In [8]:
def get_data(name):
    file_name=f"{name}.csv"
    file_path=os.path.join(PROJ_DIR,DATA_DIR,file_name)
    return pd.read_csv(file_path)
    

SyntaxError: unterminated string literal (detected at line 1) (3698387336.py, line 1)

In [11]:
flights=get_data("flight_price")


In [12]:
flights


Airline Date_of_Journey    Source Destination  \
0           IndiGo      24/03/2019  Banglore   New Delhi   
1        Air India       1/05/2019   Kolkata    Banglore   
2      Jet Airways       9/06/2019     Delhi      Cochin   
3           IndiGo      12/05/2019   Kolkata    Banglore   
4           IndiGo      01/03/2019  Banglore   New Delhi   
...            ...             ...       ...         ...   
10678     Air Asia       9/04/2019   Kolkata    Banglore   
10679    Air India      27/04/2019   Kolkata    Banglore   
10680  Jet Airways      27/04/2019  Banglore       Delhi   
10681      Vistara      01/03/2019  Banglore   New Delhi   
10682    Air India       9/05/2019     Delhi      Cochin   

                       Route Dep_Time  Arrival_Time Duration Total_Stops  \
0                  BLR → DEL    22:20  01:10 22 Mar   2h 50m    non-stop   
1      CCU → IXR → BBI → BLR    05:50         13:15   7h 25m     2 stops   
2      DEL → LKO → BOM → COK    09:25  04:25 10 Jun      19h     2 stops   
3            CCU → NAG → BLR    18:05         23:30   5h 25m      1 stop   
4            BLR → NAG → DEL    16:50         21:35   4h 45m      1 stop   
...                      ...      ...           ...      ...         ...   
10678              CCU → BLR    19:55         22:25   2h 30m    non-stop   
10679              CCU → BLR    20:45         23:20   2h 35m    non-stop   
10680              BLR → DEL    08:20         11:20       3h    non-stop   
10681              BLR → DEL    11:30         14:10   2h 40m    non-stop   
10682  DEL → GOI → BOM → COK    10:55         19:15   8h 20m     2 stops   

      Additional_Info  Price  
0             No info   3897  
1             No info   7662  
2             No info  13882  
3             No info   6218  
4             No info  13302  
...               ...    ...  
10678         No info   4107  
10679         No info   4145  
10680         No info   7229  
10681         No info  12648  
10682         No info  11753  

[10683 rows x 11 columns]

In [13]:
flights.sample(50
              )

Airline Date_of_Journey    Source Destination  \
2893            SpiceJet      24/05/2019   Kolkata    Banglore   
6012         Jet Airways      12/05/2019   Kolkata    Banglore   
180          Jet Airways       1/06/2019     Delhi      Cochin   
4066            SpiceJet      21/03/2019    Mumbai   Hyderabad   
4911              IndiGo      09/03/2019  Banglore   New Delhi   
7355             Vistara       9/06/2019  Banglore       Delhi   
7786              IndiGo      15/06/2019   Chennai     Kolkata   
3229         Jet Airways      21/05/2019   Kolkata    Banglore   
10455             IndiGo       1/06/2019     Delhi      Cochin   
1147              IndiGo      24/06/2019   Kolkata    Banglore   
4771           Air India       3/03/2019     Delhi      Cochin   
8209   Multiple carriers      24/05/2019     Delhi      Cochin   
684               IndiGo      27/05/2019  Banglore       Delhi   
2822           Air India      24/05/2019   Kolkata    Banglore   
7775   Multiple carriers      18/05/2019     Delhi      Cochin   
8495              IndiGo      27/04/2019  Banglore       Delhi   
4834           Air India       6/04/2019  Banglore       Delhi   
3032         Jet Airways      01/03/2019  Banglore   New Delhi   
2947   Multiple carriers       3/06/2019     Delhi      Cochin   
244              Vistara      18/03/2019   Chennai     Kolkata   
425          Jet Airways       6/06/2019     Delhi      Cochin   
2155   Multiple carriers      27/06/2019     Delhi      Cochin   
1141         Jet Airways      27/03/2019     Delhi      Cochin   
7795         Jet Airways       9/05/2019     Delhi      Cochin   
7857   Multiple carriers      27/03/2019     Delhi      Cochin   
9830              IndiGo      15/04/2019     Delhi      Cochin   
3181              IndiGo       1/05/2019  Banglore       Delhi   
7058            Air Asia       9/04/2019  Banglore       Delhi   
5946              IndiGo       6/06/2019    Mumbai   Hyderabad   
3205              IndiGo      24/06/2019    Mumbai   Hyderabad   
6938            Air Asia      12/06/2019   Kolkata    Banglore   
5336         Jet Airways      24/06/2019     Delhi      Cochin   
5490   Multiple carriers      18/05/2019     Delhi      Cochin   
7600         Jet Airways      12/05/2019   Kolkata    Banglore   
8891         Jet Airways       1/05/2019   Kolkata    Banglore   
3410         Jet Airways       1/05/2019   Kolkata    Banglore   
7587         Jet Airways       6/03/2019     Delhi      Cochin   
8396         Jet Airways      24/06/2019     Delhi      Cochin   
2497           Air India      01/03/2019  Banglore   New Delhi   
8611              IndiGo      01/03/2019  Banglore   New Delhi   
8686              IndiGo       9/06/2019   Chennai     Kolkata   
5560           Air India      24/06/2019     Delhi      Cochin   
1646         Jet Airways      06/03/2019  Banglore   New Delhi   
7695              IndiGo       9/06/2019     Delhi      Cochin   
255             SpiceJet      21/03/2019    Mumbai   Hyderabad   
5089              IndiGo      12/06/2019     Delhi      Cochin   
8000         Jet Airways      18/03/2019  Banglore   New Delhi   
7150   Multiple carriers      27/06/2019     Delhi      Cochin   
3716              IndiGo       3/06/2019   Chennai     Kolkata   
2887   Multiple carriers      18/05/2019     Delhi      Cochin   

                             Route Dep_Time  Arrival_Time Duration  \
2893                     CCU → BLR    22:20  00:40 25 May   2h 20m   
6012               CCU → BOM → BLR    21:10  10:05 13 May  12h 55m   
180                DEL → BOM → COK    10:00  04:25 02 Jun  18h 25m   
4066                     BOM → HYD    22:45  00:10 22 Mar   1h 25m   
4911               BLR → HYD → DEL    16:55         21:50   4h 55m   
7355                     BLR → DEL    11:30         14:20   2h 50m   
7786                     MAA → CCU    14:45         17:05   2h 20m   
3229               CCU → BOM → BLR    21:10  22:35 22 May  25h 25m   
10455               

In [14]:
flights.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10683 non-null  object
 1   Date_of_Journey  10683 non-null  object
 2   Source           10683 non-null  object
 3   Destination      10683 non-null  object
 4   Route            10682 non-null  object
 5   Dep_Time         10683 non-null  object
 6   Arrival_Time     10683 non-null  object
 7   Duration         10683 non-null  object
 8   Total_Stops      10682 non-null  object
 9   Additional_Info  10683 non-null  object
 10  Price            10683 non-null  int64 
dtypes: int64(1), object(10)
memory usage: 918.2+ KB


-  The data set contains 10,683 rows and 11 features
-   column "route and total stop " have missing values
-   The data type of some features are not appropriate




### Prelimnary Analysis


#### 1]check data Type


In [31]:
flights.dtypes

Airline            object
Date_of_Journey    object
Source             object
Destination        object
Route              object
Dep_Time           object
Arrival_Time       object
Duration           object
Total_Stops        object
Additional_Info    object
Price               int64
dtype: object

In [33]:
value=flights.Dep_Time.iloc[6]
value

'18:55'

In [25]:
flights.head()`

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897  
1    05:50         13:15   7h 25m     2 stops         No info   7662  
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882  
3    18:05         23:30   5h 25m      1 stop         No info   6218  
4    16:50         21:35   4h 45m      1 stop         No info  13302

#### 2]Check For Duplicates

In [35]:
flights.duplicated().sum()

220

In [38]:
(
    flights.loc[flights.duplicated(keep=False)].sort_values(["Airline","Date_of_Journey","Source","Destination"])
)

Airline Date_of_Journey    Source Destination                  Route  \
6321  Air India      01/03/2019  Banglore   New Delhi  BLR → BOM → AMD → DEL   
9848  Air India      01/03/2019  Banglore   New Delhi  BLR → BOM → AMD → DEL   
572   Air India      03/03/2019  Banglore   New Delhi              BLR → DEL   
8168  Air India      03/03/2019  Banglore   New Delhi              BLR → DEL   
1495  Air India       1/04/2019   Kolkata    Banglore  CCU → DEL → COK → BLR   
...         ...             ...       ...         ...                    ...   
2692   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
2870   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
3711   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
2634    Vistara      24/03/2019  Banglore   New Delhi              BLR → DEL   
5470    Vistara      24/03/2019  Banglore   New Delhi              BLR → DEL   

     Dep_Time  Arrival_Time Duration Total_Stops  \
6321    08:50  23:55 02 Mar   39h 5m     2 stops   
9848    08:50  23:55 02 Mar   39h 5m     2 stops   
572     21:10         23:55   2h 45m    non-stop   
8168    21:10         23:55   2h 45m    non-stop   
1495    10:00  01:20 02 Apr  15h 20m     2 stops   
...       ...           ...      ...         ...   
2692    05:45         08:35   2h 50m    non-stop   
2870    05:45         08:35   2h 50m    non-stop   
3711    20:30         23:20   2h 50m    non-stop   
2634    11:30         14:10   2h 40m    non-stop   
5470    11:30         14:10   2h 40m    non-stop   

                   Additional_Info  Price  
6321                       No info  17135  
9848                       No info  17135  
572                        No info   7591  
8168                       No info   7591  
1495                       No info  10408  
...                            ...    ...  
2692  No check-in baggage included   4273  
2870  No check-in baggage included   4273  
3711  No check-in baggage included   3873  
2634                       No info   5403  
5470                       No info   5403  

[415 rows x 11 columns]

#### oberavations from premlinary Analyis

 -
  The data type of "Date of journy" ,"Dep_time "and "Arival_time" should be  change to Dtaetime
 - The type of "Duration" and "Stops " is mixed .its should be numeric type
 - There are 220  duplicates .These Should be Removed
   

## 4 Detail Analysis

In [67]:
flights.select_dtypes(include="O").columns

Index(['Airline', 'Date_of_Journey', 'Source', 'Destination', 'Route',
       'Dep_Time', 'Arrival_Time', 'Duration', 'Total_Stops',
       'Additional_Info'],
      dtype='object')

-
  i want to remove the extra spaces from strip  how will i will i do

-  flights["airline"]=flights["airline"].str.strip()
-  its overriding  the current dateframe
-  df_ will hold the cureent state of the object

## 5 Cleaning Operations

In [72]:
def clean_data(df):
    return (df
            .assign(**{
                col:df[col].str.strip()
                for col in df.select_dtypes(include="O").columns
                
            }
                
            )
            .rename(columns=str.lower)
          
           )
    
    
    

In [73]:
clean_data(flights)

airline date_of_journey    source destination  \
0           IndiGo      24/03/2019  Banglore   New Delhi   
1        Air India       1/05/2019   Kolkata    Banglore   
2      Jet Airways       9/06/2019     Delhi      Cochin   
3           IndiGo      12/05/2019   Kolkata    Banglore   
4           IndiGo      01/03/2019  Banglore   New Delhi   
...            ...             ...       ...         ...   
10678     Air Asia       9/04/2019   Kolkata    Banglore   
10679    Air India      27/04/2019   Kolkata    Banglore   
10680  Jet Airways      27/04/2019  Banglore       Delhi   
10681      Vistara      01/03/2019  Banglore   New Delhi   
10682    Air India       9/05/2019     Delhi      Cochin   

                       route dep_time  arrival_time duration total_stops  \
0                  BLR → DEL    22:20  01:10 22 Mar   2h 50m    non-stop   
1      CCU → IXR → BBI → BLR    05:50         13:15   7h 25m     2 stops   
2      DEL → LKO → BOM → COK    09:25  04:25 10 Jun      19h     2 stops   
3            CCU → NAG → BLR    18:05         23:30   5h 25m      1 stop   
4            BLR → NAG → DEL    16:50         21:35   4h 45m      1 stop   
...                      ...      ...           ...      ...         ...   
10678              CCU → BLR    19:55         22:25   2h 30m    non-stop   
10679              CCU → BLR    20:45         23:20   2h 35m    non-stop   
10680              BLR → DEL    08:20         11:20       3h    non-stop   
10681              BLR → DEL    11:30         14:10   2h 40m    non-stop   
10682  DEL → GOI → BOM → COK    10:55         19:15   8h 20m     2 stops   

      additional_info  price  
0             No info   3897  
1             No info   7662  
2             No info  13882  
3             No info   6218  
4             No info  13302  
...               ...    ...  
10678         No info   4107  
10679         No info   4145  
10680         No info   7229  
10681         No info  12648  
10682         No info  11753  

[10683 rows x 11 columns]

## 6 Split The Data